dask/base.py

```diff
@@ -982,7 +982,7 @@ def _normalize_function(func):
         try:
             import cloudpickle
 
-            return cloudpickle.dumps(func, protocol=0)
+            return cloudpickle.dumps(func, protocol=None)
```

Comment:
We want to make sure the version of pickle used for tokenizing is identical across all versions of python that we care about. In this case specifying `protocol=None` means to use the default version of pickle for the provided python, which can change in a new release (creating an incompatibility). I'd rather we hardcode protocol 4 (supported for python >= 3.4) here so that if a new python release makes protocol 5 the default things don't implicitly change.

Reply:
That makes sense, thanks for the explanation Jim! I'll change this and also add in a test per @ian-r-rose's suggestion (sorry for the delay on that!)

Reply:
From reading [this thread](https://github.com/dask/dask/pull/8557#discussion_r794749520) on #8557 it seems this should be changed to `protocol=4` in both `cloudpickle.dumps` calls

Reply:
actually, I think this [test](https://github.com/dask/dask/pull/8557/files#diff-f0e433eb1ab15f20457531618797625296d28d127627463707f5274fcf24510e) Gabe added addresses the original issue

Reply:
@scharlottej13 does it? I had thought it wasn't quite the same thing https://github.com/dask/dask/pull/8557#issuecomment-1010576643

Reply:
@gjoseph92 the test I was going to add was to create a new DataClass and then call delayed on a function in that class, but really that'd be testing if it's tokenizing correctly, right? I guess the part that might still be missing is testing with `pure=True`, ie this snippet with `pure=False` works with the current `protocol=0`, but doesn't with `pure=True`:
```python
from dask import delayed
from dataclasses import dataclass, field


@dataclass(frozen=True)
class A:
    param: float = field(repr=False)

    def get_param(self):
        return self.param

    def get_delayed_param(self, *args, **kwargs):
        return delayed(self.get_param)(*args, **kwargs)

(A(1).get_delayed_param() - A(0).get_delayed_param()).compute()
```

Reply:
summary from chat w/ @gjoseph92-- it does seem that the test in #8557 will also test the underlying reason for the original issue (tokenizing DataClasses), but I'll add in a simple test that more explicitly addresses the original issue since there is the additional layer of having a bound method.